packages/next/src/server/app-render/staged-rendering.ts

```diff
@@ -32,23 +44,152 @@ export class StagedRenderingController {
         },
         { once: true }
       )
+
+      this.mayAbandon = true
+    }
+  }
+
+  onStage(stage: NonStaticRenderStage, callback: () => void) {
+    if (this.currentStage >= stage) {
+      callback()
+    } else if (stage === RenderStage.Runtime) {
+      this.runtimeStageListeners.push(callback)
+    } else if (stage === RenderStage.Dynamic) {
+      this.dynamicStageListeners.push(callback)
+    } else {
+      // This should never happen
+      throw new InvariantError(`Invalid render stage: ${stage}`)
+    }
+  }
+
+  canInterrupt() {
+    const boundaryStage = this.hasRuntimePrefetch
+      ? RenderStage.Dynamic
+      : RenderStage.Runtime
+    return this.currentStage < boundaryStage
+  }
+
+  interruptCurrentStageWithReason(reason: Error) {
+    if (this.mayAbandon) {
+      return this.abandonRenderImpl()
+    } else {
+      switch (this.currentStage) {
+        case RenderStage.Static: {
+          // We cannot abandon this render. We need to advance to the Dynamic phase
+          // but we must also capture the interruption reason.
+          this.currentStage = RenderStage.Dynamic
+          this.staticInterruptReason = reason
+
+          const runtimeListeners = this.runtimeStageListeners
+          for (let i = 0; i < runtimeListeners.length; i++) {
+            runtimeListeners[i]()
+          }
+          runtimeListeners.length = 0
+          this.runtimeStagePromise.resolve()
+
+          const dynamicListeners = this.dynamicStageListeners
+          for (let i = 0; i < dynamicListeners.length; i++) {
+            dynamicListeners[i]()
+          }
+          dynamicListeners.length = 0
+          this.dynamicStagePromise.resolve()
```

Comment:
isn't this just an inlined impl of `this.advanceStage(RenderStage.Dynamic)`? not sure why we need to duplicate that here
```suggestion
          this.staticInterruptReason = reason
		  this.advanceStage(RenderStage.Dynamic)
```